### ISSUE TITLE
$mdInkRipple needs an animation complete event

### ISSUE THREAD

**Issue by peffreycollins** [angular/material#7196]
It would be nice if $mdInkRipple had an onAnimationComplete() method, similar to the Polymer <paper-ripple> element's onAnimationComplete().

https://elements.polymer-project.org/elements/paper-ripple#method-onAnimationComplete

**Comment by DevVersion** [comment]
@peffreycollins Do you want that as an element attribute or just a method in the ripple service? I think an attribute will be also good. Please tell me more about your use-case.

**Comment by peffreycollins** [comment]
Good question.  It would be nice to have it as an element attribute - 
<md-button md-on-animation-complete="myFunction()">

Having said that, I possibly just need to optimize my application better for performance, but maybe there should be a separate performance optimization discussion on this for Angular Material apps using ripple effects in their navigation menus similar to Angular material.   

I am using an left-hand icon navigation menu with polymer's <paper-ripple> paper element to display the ripple animation onclick of each nav menu item (this was in place before I integrated with angular material -- looking to replace it).  I am doing some pretty heavy lifting (loading a number of dynamic widget templates asynchronously each with their own respective REST calls).  Depending on the number of widgets on the screen, this can cause a stutter in the ripple effect immediately when navigating to the new view.  In my case, I wasn't transitioning my ui-router state until the "onAnimationComplete" event was triggered in polymer, which provided me with a few milliseconds of grace period.  Obviously, the downside to this approach is that no effort is made to change routes until the animation has completed, but the animation is so quick that it didn't matter and increased user experience.    

To emulate this animation stutter when navigating on the angular material site,  navigate to the Autocomplete (https://material.angularjs.org/latest/demo/autocomplete) from the navigation menu to see the slight stutter in the ripple.

**Comment by DevVersion** [comment]
@peffreycollins Thank's for the detailed information. I'll dive into this tomorrow. 

To your use-case, I exactly know what you mean, I've noticed that really often too. 
An even better example for that `stutter` in the ripple, is at https://angular.io/ (the page is even soo fast, that the ripples can't complete)

**Comment by peffreycollins** [comment]
@DevVersion hey -- thanks a bunch!  :) Glad I'm not the only one noticing this.  

I wonder how the devs of the [Google Design page](http://design.google.com/) are handling this with their animations in general.  It looks like they are delaying certain things cleverly as well to make appear to be very smooth.  Their route changes immediately, but they aren't completing their animation and updating the view until everything is ready.

Either way, I think the on-animation-complete could provide a nice hook for devs like Polymer does.  

Thanks again.

**Comment by peffreycollins** [comment]
An alternative approach to this might be to have the user set the ripple duration in ms, and then the dev could use this value to set a $timeout().

`<md-button md-ripple-duration="1000">`

```
$timeout(function(){
//Some action here}, 1000)
```

Not sure which one your team would prefer...just throwing it out there.  Maybe both could be useful!

**Comment by DevVersion** [comment]
I don't know, but the `md-ripple-duration` seems for me to hard-coded. I think the complete event will be a really smart solution and can be useful for much more things, than just a navigation.

**Comment by peffreycollins** [comment]
Agreed.  Thanks again :) .

**Comment by DevVersion** [comment]
@peffreycollins I recently talked with the team about that feature request. We decided to not implement that event, because there are not many use-cases. To the case with the `routing` / `stuttering` when changing the route is caused by a performance issue, and we all agreed, that delaying the route change is not recommend. 

And in addition to that, it seems that the ripple effect is currently running in the `Main Thread` and is not taking advantage of `requestAnimationFrame`, `Off Threading` and GPU Rendering.

Here is a related performance PR: #6631 
And here is a little blog about the `Main Thread` and `Off Threading` 
- http://www.phpied.com/css-animations-off-the-ui-thread/
